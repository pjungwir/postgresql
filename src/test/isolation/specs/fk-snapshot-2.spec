# RI Trigger test
#
# Test C-based referential integrity enforcement.
# Under REPEATABLE READ we need some snapshot trickery in C,
# or we would permit things that violate referential integrity.

setup
{
  CREATE TABLE parent (parent_id SERIAL NOT NULL PRIMARY KEY);
  CREATE TABLE child (
	child_id SERIAL NOT NULL PRIMARY KEY,
	parent_id INTEGER REFERENCES parent);
  INSERT INTO parent VALUES(1);
}

teardown { DROP TABLE parent, child; }

session s1
step s1rc	{ BEGIN ISOLATION LEVEL READ COMMITTED; }
step s1rr	{ BEGIN ISOLATION LEVEL REPEATABLE READ; }
step s1ser	{ BEGIN ISOLATION LEVEL SERIALIZABLE; }
step s1del	{ DELETE FROM parent WHERE parent_id = 1; }
step s1c	{ COMMIT; }

session s2
step s2rc	{ BEGIN ISOLATION LEVEL READ COMMITTED; }
step s2rr	{ BEGIN ISOLATION LEVEL REPEATABLE READ; }
step s2ser	{ BEGIN ISOLATION LEVEL SERIALIZABLE; }
step s2ins	{ INSERT INTO child VALUES (1, 1); }
step s2c	{ COMMIT; }

# Violates referential integrity unless we use a crosscheck snapshot,
# which is up-to-date compared with the transaction's snapshot.
permutation s1rr s2rr s2ins s1del s2c s1c

# Raises a can't-serialize exception
# when the INSERT trigger does SELECT FOR KEY SHARE:
permutation s1rr s2rr s1del s2ins s1c s2c

# Test the same scenarios in READ COMMITTED:
# A crosscheck snapshot is not required here.
permutation s1rc s2rc s2ins s1del s2c s1c
permutation s1rc s2rc s1del s2ins s1c s2c

# Test the same scenarios in SERIALIZABLE:
# We should report the FK violation:
permutation s1ser s2ser s2ins s1del s2c s1c
# We raise a concurrent update error
# which is good enough:
permutation s1ser s2ser s1del s2ins s1c s2c
