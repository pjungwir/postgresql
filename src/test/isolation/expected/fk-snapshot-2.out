Parsed test spec with 2 sessions

starting permutation: s1rr s2rr s2ins s1del s2c s1c
step s1rr: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2rr: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2ins: INSERT INTO child VALUES (1, 1);
step s1del: DELETE FROM parent WHERE parent_id = 1; <waiting ...>
step s2c: COMMIT;
step s1del: <... completed>
ERROR:  update or delete on table "parent" violates foreign key constraint "child_parent_id_fkey" on table "child"
step s1c: COMMIT;

starting permutation: s1rr s2rr s1del s2ins s1c s2c
step s1rr: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2rr: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1del: DELETE FROM parent WHERE parent_id = 1;
step s2ins: INSERT INTO child VALUES (1, 1); <waiting ...>
step s1c: COMMIT;
step s2ins: <... completed>
ERROR:  could not serialize access due to concurrent update
step s2c: COMMIT;

starting permutation: s1rc s2rc s2ins s1del s2c s1c
step s1rc: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2rc: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2ins: INSERT INTO child VALUES (1, 1);
step s1del: DELETE FROM parent WHERE parent_id = 1; <waiting ...>
step s2c: COMMIT;
step s1del: <... completed>
ERROR:  update or delete on table "parent" violates foreign key constraint "child_parent_id_fkey" on table "child"
step s1c: COMMIT;

starting permutation: s1rc s2rc s1del s2ins s1c s2c
step s1rc: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2rc: BEGIN ISOLATION LEVEL READ COMMITTED;
step s1del: DELETE FROM parent WHERE parent_id = 1;
step s2ins: INSERT INTO child VALUES (1, 1); <waiting ...>
step s1c: COMMIT;
step s2ins: <... completed>
ERROR:  insert or update on table "child" violates foreign key constraint "child_parent_id_fkey"
step s2c: COMMIT;

starting permutation: s1ser s2ser s2ins s1del s2c s1c
step s1ser: BEGIN ISOLATION LEVEL SERIALIZABLE;
step s2ser: BEGIN ISOLATION LEVEL SERIALIZABLE;
step s2ins: INSERT INTO child VALUES (1, 1);
step s1del: DELETE FROM parent WHERE parent_id = 1; <waiting ...>
step s2c: COMMIT;
step s1del: <... completed>
ERROR:  update or delete on table "parent" violates foreign key constraint "child_parent_id_fkey" on table "child"
step s1c: COMMIT;

starting permutation: s1ser s2ser s1del s2ins s1c s2c
step s1ser: BEGIN ISOLATION LEVEL SERIALIZABLE;
step s2ser: BEGIN ISOLATION LEVEL SERIALIZABLE;
step s1del: DELETE FROM parent WHERE parent_id = 1;
step s2ins: INSERT INTO child VALUES (1, 1); <waiting ...>
step s1c: COMMIT;
step s2ins: <... completed>
ERROR:  could not serialize access due to concurrent update
step s2c: COMMIT;
